lib/puma/systemd.rb

```diff
@@ -0,0 +1,47 @@
+# frozen_string_literal: true
+
+require 'sd_notify'
+
+module Puma
+  class Systemd
+    def initialize(events)
+      @events = events
+    end
+
+    def hook_events
+      @events.on_booted { SdNotify.ready }
+      @events.on_stopped { SdNotify.stopping }
+      @events.on_restart { SdNotify.reloading }
+    end
+
+    def start_watchdog
+      ping_f = watchdog_sleep_time
+      return unless ping_f
+
+      log "Pinging systemd watchdog every #{ping_f.round(1)} sec"
+      Thread.new do
+        loop do
+          sleep ping_f
+          SdNotify.watchdog
+        end
+      end
+    end
+
+    private
+
+    def watchdog_sleep_time
+      return unless SdNotify.watchdog?
```

Comment:
I think it would improve the understanding here to move this up to just after the `def start_watchdog` line.

Reply:
Agreed, moved.